update mixin, multiple additions

AutoSponge opened the pull request:
adding first
adding reduceRight
adding some
adding every
adding findIndex
adding size
adding Object.assign
adding now
adding templates

AutoSponge committed:
adding first
adding reduceRight
adding some
adding every
adding findIndex
adding size
adding Object.assign
adding now
adding templates

fson commented on README.md:
````diff
@@ -1,5 +1,18 @@
+  ```javascript
+  const [n] = [5, 4, 3, 2, 1];
+  ```
+
````
I would use just `array[0]` as the "vanilla" example, but you probably wouldn't use `_.first` in a simple case like that anyway. (It's more useful for function composition.)

These are also not equivalent, the second one assigns it to a constant, while the first one is just the expression.

AutoSponge replied:
Yeah, I agree. Do you think it's worth including `const first = ([n]) => n;` as a simple definition of `first`?

fson replied:
Maybe it's not the best example. The point I tried to get across in the blog post is that you can often just write

``` javascript
array.map(x => x[0])
```

instead of

``` javascript
_.map(array, _.first)
```

If we define our own `first` function, people will come to tell we're now implementing an untested, ad hoc implementation of half of Underscore. :wink:

fson commented on README.md:
````diff
@@ -28,6 +41,48 @@
+  array.reduceRight(iteratee)
+  ```
+
+#### Every
````
The heading should be a short description of what the example does, e.g. "Test if all elements in the array pass the predicate"

fson commented on README.md:
````diff
@@ -126,6 +181,20 @@
   array.indexOf(value)
   ```
 
+#### Find the index in an array by predicate
````
:+1:

fson commented on README.md:
````diff
@@ -156,6 +225,20 @@
   Object.keys(object)
   ```
 
+#### Size
````
I think "Number of keys in an object" would be a better title.

fson commented on README.md:
````diff
@@ -206,6 +289,12 @@
+* ES2015
+
+  ```javascript
+  Object.assign({}, source, { a: false });
````
:+1:

fson commented on README.md:
````diff
@@ -331,3 +420,33 @@
+
+  ```javascript
+  Date.now()
+  ```
````
:+1:

fson commented on README.md:
````diff
@@ -331,3 +420,33 @@
+
+  ```javascript
+  _.template("hello: <%= name %>");
+  compiled({name: 'moe'});
````
The variable `compiled` is currently not defined. I think this should be:

``` javascript
var greeting = _.template("hello <%= name %>");
greeting({ name: 'moe' });
```

To make the examples match more closely, maybe we could create a template function in the ES2015 example as well, e.g.

``` javascript
const greeting = ({ name }) => `hello ${name}`;
greeting({ name: 'moe' });
```

fson commented on README.md:
````diff
@@ -28,6 +41,48 @@
+
+  ```javascript
+  array.reduceRight(iteratee)
+  ```
````
If we include `reduceRight`, maybe we should also include `reduce`?

AutoSponge committed:
respond to comments
first is now a function
title change for size
title change for reduceright

AutoSponge committed:
respond to comments
adding reduce
adding "memo" param to reduce and reduceRight for consistency with underscore api docs

fson commented on README.md:
````diff
@@ -28,6 +42,63 @@
   array.map(iteratee)
   ```
 
+#### Use a predicate to accumulate a single value from an array (left-to-right)
````
I understand "predicate" to mean a function with a boolean return value. Just "function" would be better. Otherwise looks good.

fson commented on README.md:
````diff
@@ -1,5 +1,19 @@
+
+  ```javascript
+  const first = ([n]) => n;
+  first([5, 4, 3, 2, 1]);
````
I'm still not sure about this. I wouldn't define a `first` function, because it might be misleading – the point is that you might not need a `first` function.

If the idea here is to show how you can use destructuring, I think maybe we could do this:

``` javascript
var first = _.first(array);
```

and

``` javascript
var [first] = array; 
```

AutoSponge replied:
That's what it was before... I'm just going to delete it.

AutoSponge committed:
responding to comments
removing first
changing wording of reduce[right] to function instead of predicate

fson commented:
@AutoSponge Looks good. Can you merge from master or rebase so this can be merged. Thanks!

fson commented:
Please squash everything into one commit.

AutoSponge committed:
adding reduceRight
adding some
adding every
adding findIndex
adding size
adding Object.assign
adding now
adding templates

AutoSponge committed:
adding reduceRight
adding some
adding every
adding findIndex
adding size
adding Object.assign
adding now
adding templates
